### PR TITLE
DSV4 MTP Support

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Adds native architecture support for the Multi-Token Prediction (MTP) step used by DeepSeek-V4.

Previously, DeepSeek-V4 models were unsupported with `mtp_num_layers > 0` because DeepSeek-V4 maintains a different hidden state projection topology compared to standard architectures. DeepSeek actively relies on an expanded 4-dimensional hidden tensor context (`[batch, seq, hc_mult, dim]`) that natively processes hyper-connections inside the MTP step, rather than operating sequentially on a pre-squashed 3D intermediate backbone state.

This PR aligns MaxText's `MultiTokenPredictionBlock` with the reference DeepSeek-V4 PyTorch architecture by refactoring the internal state propagation loop.
Specifically, it:
- Skips the explicit 4D -> 3D reduction step (`self.hc_head`) previously executed by default in `NNXDecoder.__call__` and moves it explicitly back into `apply_output_head` via an intercepted `reduce_mhc=True` toggle.
- Preserves the full unabridged 4D state and broadcasts positional embeddings accordingly within the MTP layer to process sequential tokens while fully preserving hyper-connection topologies across consecutive iterations.
- Automatically initializes standalone instances of `DeepSeek4HyperHead` explicitly for independent MTP blocks so that 4D reduction only occurs safely when projecting the final sequence into next-token prediction logits.

# Tests

- Configured and executed a local CPU Mock pipeline (`run_mock.py`) initialized with `deepseek4-tiny.yml` configured to `base_num_decoder_layers: 5` and `mtp_num_layers: 1` to ensure exact mesh parity and identical context-rolling structures against the upstream DeepSeek PyTorch reference.
- Passed all generic architecture tests in `tests/unit/multi_token_prediction_test.py` via `pytest`, modifying test mocks seamlessly to conform to the new `reduce_mhc` parameters.
- Deleted the explicit DeepSeek exception from `tests/unit/flop_calculation_test.py` and validated that FLOPs automatically scale to correctly encompass the DeepSeek MTP routing parameters. 
- Passed all remaining distributed gradient accumulation sweeps (`tests/unit/gradient_accumulation_nnx_test.py`, `tests/unit/train_nnx_test.py`, and `tests/unit/deepseek4_mtp_test.py`).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/multi_token_prediction.py [hunk 1]
```diff
@@ -108,12 +108,30 @@ def __init__(
...
-        config=cfg,
-        mesh=mesh,
-        model_mode=MODEL_MODE_TRAIN,
-        rngs=rngs,
-    )
+    if cfg.decoder_block == DecoderBlockType.DEEPSEEK4:
+      from maxtext.models import deepseek4  # pylint: disable=import-outside-toplevel
+      from maxtext.layers import mhc  # pylint: disable=import-outside-toplevel
+
+      self.transformer_layer = deepseek4.DeepSeek4LayerToLinen(
```
Comment (JamesDeng42) [inline src/maxtext/layers/multi_token_prediction.py]: please use DeepSeek4DecoderLayer for nnx module

File: src/maxtext/layers/nnx_decoders.py [hunk 2]
```diff
@@ -1406,10 +1386,18 @@ def _apply_embedding(
...
-  def apply_output_head(self, shared_embedding, y, deterministic, model_mode):
+  def apply_output_head(self, shared_embedding, y, deterministic, model_mode, reduce_mhc=True):
     """Applies final normalization and projects hidden states to logits."""
 
     cfg = self.config
+    if reduce_mhc and cfg.mhc_expansion_rate > 1 and y.ndim >= 4:
+      if cfg.decoder_block == DecoderBlockType.DEEPSEEK4:
+        y = self.hc_head(y)
+      else:
+
```
Comment (JamesDeng42) [inline src/maxtext/layers/nnx_decoders.py]: nit: remove this blank line

File: src/maxtext/layers/nnx_decoders.py
Comment (JamesDeng42) [inline src/maxtext/layers/nnx_decoders.py:1957]: if we remove this block of code, this change is not scoped to DSV4, is this intended?

File: tests/unit/deepseek4_mtp_test.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+"""Unit tests for DeepSeek V4 MTP execution logic."""
+
+import unittest
+import jax
```
Comment (JamesDeng42) [inline tests/unit/deepseek4_mtp_test.py:5]: Could we make this a forward/execution test rather than an initialization-only test? the unit test is too weak for this change.

File: src/maxtext/layers/multi_token_prediction.py [hunk 4]
```diff
@@ -319,7 +342,12 @@ def __call__(
           model_mode=self.decoder.model_mode,
       )
 
-      mtp_logits = self.decoder.apply_output_head(shared_embedding, mtp_hidden_state, deterministic, model_mode)
+      mtp_hidden_state_to_head = mtp_hidden_state
+      if cfg.mhc_expansion_rate > 1 and cfg.decoder_block == DecoderBlockType.DEEPSEEK4:
+        mtp_hidden_state_to_head = mtp_layer.hc_head(mtp_hidden_state)
+      mtp_logits = self.decoder.apply_output_head(
```
Comment (JamesDeng42) [inline src/maxtext/layers/multi_token_prediction.py:368]: DSV4 MTP should use an MTP-local final norm here rather than the backbone decoder_norm ? could you double check the official implementation and make sure this should match the official one.

**Review by JamesDeng42 — COMMENTED** [review]
please address my comments and gemini's comments.

File: src/maxtext/layers/multi_token_prediction.py [hunk 5]
```diff
@@ -187,6 +205,11 @@ def __call__(
       hidden_state_pspec = jax.sharding.NamedSharding(self.mesh, jax.typeof(hidden_state_norm).sharding.spec)
       embedding_norm = jax.reshard(embedding_norm, hidden_state_pspec)
 
+    if self.config.decoder_block == DecoderBlockType.DEEPSEEK4 and self.config.mhc_expansion_rate > 1:
+      # e: (batch, seq, dim), x: (batch, seq, hc, dim)
+      # We need to broadcast e to (batch, seq, hc, dim) before concatenation
+      mhc = self.config.mhc_expansion_rate
+      embedding_norm = jnp.expand_dims(embedding_norm, axis=2).repeat(mhc, axis=2)
```
[the remark this answers is not in the document]
Comment (vlad-karp) [reply src/maxtext/layers/multi_token_prediction.py:232]: Seems to be a mislocated comment
